File: iOS/TUILiveKit/Sources/LiveStream/Main/Manager/LiveStreamManager.swift
//
//  LiveStreamManager.swift
//  TUILiveKit
//
//  Created by jeremiawang on 2024/11/18.
//

import Foundation
import Combine
import RTCCommon
import RTCRoomEngine
import LiveStreamCore
import TUILiveComponent

typealias LSRoomStateUpdateClosure = (inout LSRoomState) -> Void
typealias LSUserStateUpdateClosure = (inout LSUserState) -> Void
typealias LSMediaStateUpdateClosure = (inout LSMediaState) -> Void
typealias LSCoGuestStateUpdateClosure = (inout LSCoGuestState) -> Void

typealias InternalErrorBlock = (_ error: InternalError) -> Void

protocol LiveStreamManagerProvider: NSObject {
    func getCoreViewState<T: State>() -> T
    func subscribeCoreViewState<State, Value>(_ selector: StateSelector<State, Value>) -> AnyPublisher<Value, Never>
}

class LiveStreamManager {
    
    public let toastSubject = PassthroughSubject<String, Never>()
    public let likeSubject = PassthroughSubject<Void, Never>()
    public let floatWindowSubject = PassthroughSubject<Void, Never>()
    public let kickedOutSubject = PassthroughSubject<Void, Never>()
    
    class Context {
        let service = LSRoomEngineService()
        let coHostService = LSCoHostServiceImpl()
        weak var provider: LiveStreamManagerProvider?
        
        private(set) lazy var roomManager = LSRoomManager(context: self)
        private(set) lazy var userManager = LSUserManager(context: self)
        private(set) lazy var mediaManager = LSMediaManager(context: self)
        private(set) lazy var coGuestManager = LSCoGuestManager(context: self)
        
        private(set) lazy var coHostManager = LSCoHostManager(context: self)
        private(set) lazy var battleManager = LSBattleManager(context: self)
        
        private(set) lazy var engineObserver = LSRoomEngineObserver(context: self)
        private(set) lazy var liveListObserver = LSLiveListObserver(context: self)
        private(set) lazy var coHostObserver = LSCoHostObserver(context: self)
        
        let toastSubject: PassthroughSubject<String, Never>
        let kickedOutSubject: PassthroughSubject<Void, Never>
        
        init(provider: LiveStreamManagerProvider, toastSubject: PassthroughSubject<String, Never>, kickedOutSubject: PassthroughSubject<Void, Never>) {
            self.toastSubject = toastSubject
            self.kickedOutSubject = kickedOutSubject
            self.provider = provider
            service.addEngineObserver(engineObserver)
            service.addLiveListManagerObserver(liveListObserver)
            coHostService.addConnectionObserver(coHostObserver)
        }
        
        deinit {
            service.removeEngineObserver(engineObserver)
            service.removeLiveListManagerObserver(liveListObserver)
            coHostService.removeConnectionObserver(coHostObserver)
        }
    }
    
    private let context: Context
    private var cancellableSet: Set<AnyCancellable> = []
    
    init(provider: LiveStreamManagerProvider) {
        self.context = Context(provider: provider, toastSubject: toastSubject, kickedOutSubject: kickedOutSubject)
    }
}

// MARK: - Common
extension LiveStreamManager {
    func onError(_ error: InternalError) {
        toastSubject.send(error.localizedMessage)
    }
    
    func onCameraOpened(localVideoView: UIView) {
        context.mediaManager.onCameraOpened()
        context.mediaManager.setLocalVideoView(view: localVideoView)
    }
    
    func onReceiveGift(price: Int, senderUserId: String) {
        context.roomManager.onReceiveGift(price: price, senderUserId: senderUserId)
    }
}

// MARK: - Anchor
extension LiveStreamManager {
    func prepareLiveInfoBeforeEnterRoom(liveInfo: TUILiveInfo) {
        context.roomManager.prepareLiveInfoBeforeEnterRoom(liveInfo: liveInfo)
    }
    
    func prepareRoomIdBeforeEnterRoom(roomId: String) {
        context.roomManager.prepareRoomIdBeforeEnterRoom(roomId: roomId)
    }
    
    func onSetRoomName(_ name: String) {
        context.roomManager.onSetRoomName(name)
    }
    
    func onSetRoomPrivacy(_ mode: LiveStreamPrivacyStatus) {
        context.roomManager.onSetRoomPrivacy(mode)
    }
    
    func onSetRoomCoverUrl(_ url: String) {
        context.roomManager.onSetRoomCoverUrl(url)
    }
    
    func onStartLive(isJoinSelf: Bool, roomInfo: TUIRoomInfo) {
        context.roomManager.onStartLive(isJoinSelf: isJoinSelf, roomInfo: roomInfo)
        context.userManager.onStartLive()
    }
    
    func onStopLive() {
        context.roomManager.onStopLive()
    }
    
    func onCoHostConnectUserChanged(connectUserList: [TUIConnectionUser]) {
        context.mediaManager.changeVideoEncParams(encType: connectUserList.count > 1 ? .small : .big)
    }
    
    func onCoGuestConnectUserChanged(connectUserList: [TUIUserInfo]) {
        context.mediaManager.changeVideoEncParams(encType: connectUserList.count > 1 ? .small : .big)
    }
    
    // Cross room
    func onCrossRoomConnectionTerminated() {
        context.coHostManager.onCrossRoomConnectionTerminated()
    }
    
    // Battle
    func onRequestBattle(battleId: String, battleUserList: [TUIBattleUser]) {
        context.battleManager.onRequestBattle(battleId: battleId, battleUserList: battleUserList)
    }
    
    func onResponseBattle() {
        context.battleManager.onResponseBattle()
    }
    
    func onBattleExited() {
        context.battleManager.onBattleExited()
    }
    
    // Other
    func fetchLiveInfo(roomId: String, onSuccess: @escaping ((_ liveInfo: TUILiveInfo)->()), onError: @escaping InternalErrorBlock) {
        Task {
            do {
                let liveInfo = try await context.roomManager.fetchLiveInfo(roomId: roomId)
                DispatchQueue.main.async {
                    onSuccess(liveInfo)
                }
            } catch let err as InternalError {
                DispatchQueue.main.async {
                    onError(err)
                }
            }
        }
    }
    
    func getUserInfo(userId: String) async throws -> TUIUserInfo {
        try await context.userManager.getUserInfo(userId: userId)
    }
    
    func onDisableSendingMessageBtnClicked(userId: String, isDisable: Bool, onSuccess: @escaping TUISuccessBlock, onError: @escaping InternalErrorBlock) {
        Task {
            do {
                try await context.userManager.onDisableSendingMessageBtnClicked(userId: userId, isDisable: isDisable)
                DispatchQueue.main.async {
                    onSuccess()
                }
            } catch let err as InternalError {
                DispatchQueue.main.async {
                    onError(err)
                }
            }
        }
    }
    
    func onKickedOutBtnClicked(userId: String, onSuccess: @escaping TUISuccessBlock, onError: @escaping InternalErrorBlock) {
        Task {
            do {
                try await context.userManager.onKicedOutBtnClicked(userId: userId)
                DispatchQueue.main.async {
                    onSuccess()
                }
            } catch let err as InternalError {
                DispatchQueue.main.async {
                    onError(err)
                }
            }
        }
    }
    
    func onLockMediaStatusBtnClicked(userId: String, lockParams: TUISeatLockParams, onSuccess: @escaping TUISuccessBlock, onError: @escaping InternalErrorBlock) {
        Task {
            do {
                try await context.coGuestManager.onLockMediaStatusBtnClicked(userId: userId, lockParams: lockParams)
                DispatchQueue.main.async {
                    onSuccess()
                }
            } catch let err as InternalError {
                DispatchQueue.main.async {
                    onError(err)
                }
            }
        }
    }
}

// MARK: - Audience
extension LiveStreamManager {
    func onJoinLive(roomInfo: TUIRoomInfo) {
        context.roomManager.onJoinLive(roomInfo: roomInfo)
    }
    
    func onLeaveLive() {
        context.roomManager.onLeaveLive()
        context.userManager.onLeaveLive()
        context.mediaManager.onLeaveLive()
    }
    
    func onAudienceSliderCellInit(liveInfo: LiveInfo) {
        context.roomManager.onAudienceSliderCellInit(liveInfo: liveInfo)
    }
    
    func onStartRequestIntraRoomConnection() {
        context.coGuestManager.onStartRequestIntraRoomConnection()
    }
    
    func onRequestIntraRoomConnectionFailed() {
        context.coGuestManager.onRequestIntraRoomConnectionFailed()
    }
    
    func onStartCancelIntraRoomConnection() {
        context.coGuestManager.onStartCancelIntraRoomConnection()
    }
    
    func onCancelIntraRoomConnection() {
        context.coGuestManager.onCancelIntraRoomConnection()
    }
    
    func onCoGuestStatusChanged(status: LSCoGuestState.CoGuestStatus) {
        if status == .linking {
            context.mediaManager.changeVideoEncParams(encType: .small)
        }
    }
}

// MARK: - Observer
extension LiveStreamManager {
    func onUserConnectionRejected(userId: String) {
        toastSubject.send(.takeSeatApplicationRejected)
        context.coGuestManager.onUserConnectionRejected(userId: userId)
    }
    
    func onUserConnectionTimeout(userId: String) {
        toastSubject.send(.takeSeatApplicationTimeout)
        context.coGuestManager.onUserConnectionTimeout(userId: userId)
    }
    
    func onKickedOffSeat() {
        toastSubject.send(.kickedOutOfSeat)
        context.coGuestManager.onKickedOffSeat()
    }
    
    func onConnectionUserListChanged(list: [TUIConnectionUser]) {
        context.coHostManager.onConnectionUserListChanged(list: list)
    }
}

// MARK: - Tools
extension LiveStreamManager {
    // State
    var roomState: LSRoomState {
        context.roomManager.roomState
    }
    var mediaState: LSMediaState {
        context.mediaManager.mediaState
    }
    var userState: LSUserState {
        context.userManager.userState
    }
    var coGuestState: LSCoGuestState {
        context.coGuestManager.coGuestState
    }
    var battleState: LSBattleState {
        context.battleManager.state
    }
    var coHostState: LSCoHostState {
        context.coHostManager.state
    }
    
    // Manager
    var coHostManager: LSCoHostManager {
        context.coHostManager
    }
    var battleManager: LSBattleManager {
        context.battleManager
    }
    var mediaManager: LSMediaManager {
        context.mediaManager
    }
    
    // Other
    func getDefaultRoomName() -> String {
        return context.roomManager.getDefaultRoomName()
    }
    
    func subscribeState<State, Value>(_ selector: StateSelector<State, Value>) -> AnyPublisher<Value, Never> {
        if let sel = selector as? StateSelector<LSUserState, Value> {
            return context.userManager.subscribeState(sel)
        } else if let sel = selector as? StateSelector<LSRoomState, Value> {
            return context.roomManager.subscribeState(sel)
        } else if let sel = selector as? StateSelector<LSBattleState, Value> {
            return context.battleManager.subscribeState(sel)
        } else if let sel = selector as? StateSelector<LSCoHostState, Value> {
            return context.coHostManager.subscribeCoHostState(sel)
        } else if let sel = selector as? StateSelector<LSMediaState, Value> {
            return context.mediaManager.subscribeState(sel)
        } else if let sel = selector as? StateSelector<LSCoGuestState, Value> {
            return context.coGuestManager.subscribeState(sel)
        }
        assert(false, "Not impl")
        return Empty<Value, Never>().eraseToAnyPublisher()
    }
    
    func subscribeCoreViewState<State, Value>(_ selector: StateSelector<State, Value>) -> AnyPublisher<Value, Never> {
        guard let provider = context.provider else { return Empty<Value, Never>().eraseToAnyPublisher() }
        return provider.subscribeCoreViewState(selector)
    }
}

extension LiveStreamManager {
    var coreRoomState: RoomState {
        context.coreRoomState
    }
    var coreUserState: UserState {
        context.coreUserState
    }
    var coreMediaState: MediaState {
        context.coreMediaState
    }
    var coreCoHostState: CoHostState {
        context.coreCoHostState
    }
    var coreCoGuestState: CoGuestState {
        context.coreCoGuestState
    }
    var coreBattleState: BattleState {
        context.coreBattleState
    }
}

extension LiveStreamManager.Context {
    var coreRoomState: RoomState {
        guard let provider = provider else { return RoomState() }
        return provider.getCoreViewState()
    }
    var coreUserState: UserState {
        guard let provider = provider else { return UserState() }
        return provider.getCoreViewState()
    }
    var coreMediaState: MediaState {
        guard let provider = provider else { return MediaState() }
        return provider.getCoreViewState()
    }
    var coreCoHostState: CoHostState {
        guard let provider = provider else { return CoHostState() }
        return provider.getCoreViewState()
    }
    var coreCoGuestState: CoGuestState {
        guard let provider = provider else { return CoGuestState() }
        return provider.getCoreViewState()
    }
    var coreBattleState: BattleState {
        guard let provider = provider else { return BattleState() }
        return provider.getCoreViewState()
    }
}

extension LiveStreamManager: GiftListPanelProvider {
    func getAnchorInfo() -> GiftUser {
        let owner = coreRoomState.ownerInfo
        let giftUser = GiftUser(userId: owner.userId, name: owner.userName, avatarUrl: owner.avatarUrl)
        return giftUser
    }
}

fileprivate extension String {
    static let takeSeatApplicationRejected = internalLocalized("Take seat application has been rejected")
    static let takeSeatApplicationTimeout = internalLocalized("Take seat application timeout")
    static let kickedOutOfSeat = internalLocalized("Kicked out of seat by room owner")
}

File: Flutter/livekit/lib/live_stream/manager/live_stream_manager.dart
import 'dart:async';

import 'package:flutter/cupertino.dart';
import 'package:live_stream_core/live_core_widget/index.dart' hide CoGuestStatus;
import 'package:rtc_room_engine/rtc_room_engine.dart';
import 'package:tencent_live_uikit/live_stream/live_define.dart';
import 'package:tencent_live_uikit/live_stream/manager/module/float_window_manager.dart';
import 'package:tencent_live_uikit/live_stream/state/float_window_state.dart';
import 'package:tencent_live_uikit/live_stream/state/index.dart';

import '../../common/widget/float_window/float_window_mode.dart';
import '../api/live_stream_service.dart';
import 'module/index.dart';
import 'observer/index.dart';

typedef GetCoreState = CoreState Function();

class CoreStateProvider {
  GetCoreState getCoreState;

  CoreStateProvider({required this.getCoreState});
}

class LiveStreamManager {
  late final LiveStreamService service;

  late final StreamController<String> toastSubject;
  late final StreamController<void> likeSubject;
  late final StreamController<void> kickedOutSubject;

  late final RoomManager _roomManager;
  late final UserManager _userManager;
  late final MediaManager _mediaManager;
  late final CoGuestManager _coGuestManager;
  late final CoHostManager _coHostManager;
  late final BattleManager _battleManager;
  late final FloatWindowManager _floatWindowManager;

  late final RoomEngineObserver _roomEngineObserver;
  late final LiveListObserver _liveListObserver;
  late final LiveStreamObserver liveStreamObserver;
  late final BattleManagerObserver battleManagerObserver;
  late final LiveLayoutObserver _liveLayoutObserver;
  late final Context _context;
  late final CoreStateProvider provider;

  LiveStreamManager({required this.provider}) {
    _init();
  }

  void dispose() {
    _unInit();
  }
}

extension LiveStreamManagerWithCommon on LiveStreamManager {
  void onCameraOpened(int viewId) {
    _mediaManager.onCameraOpened();
    _mediaManager.setLocalVideoView(viewId);
  }

  void onReceiveGift(int price, String senderUserId) {
    return _roomManager.onReceiveGift(price, senderUserId);
  }

  Future<void> fetchRecommendedList({String cursor = ''}) {
    return _coHostManager.fetchRecommendedList(cursor: cursor);
  }
}

extension LiveStreamManagerWithAnchor on LiveStreamManager {
  void prepareLiveInfoBeforeEnterRoom(TUILiveInfo liveInfo) {
    _mediaManager.prepareLiveInfoBeforeEnterRoom(liveInfo);
    return _roomManager.prepareLiveInfoBeforeEnterRoom(liveInfo);
  }

  void prepareRoomIdBeforeEnterRoom(String roomId) {
    return _roomManager.prepareRoomIdBeforeEnterRoom(roomId);
  }

  void onSetRoomName(String roomName) {
    return _roomManager.onSetRoomName(roomName);
  }

  void onSetRoomPrivacy(LiveStreamPrivacyStatus mode) {
    return _roomManager.onSetRoomPrivacy(mode);
  }

  void onSetRoomCoverUrl(String url) {
    return _roomManager.onSetRoomCoverUrl(url);
  }

  void onStartPreview() {
    return _roomManager.onStartPreview();
  }

  void onStartLive(bool isJoinSelf, TUILiveInfo liveInfo) {
    return _roomManager.onStartLive(isJoinSelf, liveInfo);
  }

  void onStopLive() {
    _mediaManager.onStopLive();
    return _roomManager.onStopLive();
  }

  void onCoHostConnectUserChanged(List<TUIConnectionUser> connectedUserList) {
    // TODO: krab needs to confirm whether needs changeVideoEncParams logic here
  }

  void onCoGuestConnectUserChanged(List<TUIUserInfo> connectedUserList) {
    // TODO: krab needs to confirm whether needs changeVideoEncParams logic here
  }

  // Cross room
  void onCrossRoomConnectionTerminated() {
    _coHostManager.onCrossRoomConnectionTerminated();
  }

  void onRequestCrossRoomConnection(TUIConnectionUser user) {
    _coHostManager.onRequestConnection(user);
  }

  void onRequestCrossRoomConnectionFailed(String roomId) {
    _coHostManager.onRequestConnectionFailed(roomId);
  }

  // Battle
  void onRequestBattle(String battleId, List<TUIBattleUser> battleUserList) {
    _battleManager.onRequestBattle(battleId, battleUserList);
  }

  void onCanceledBattle() {
    _battleManager.onCanceledBattle();
  }

  void onResponseBattle() {
    _battleManager.onResponseBattle();
  }

  void onBattleExited() {
    _battleManager.onBattleExited();
  }

  bool isBattleDraw() {
    return _battleManager.isBattleDraw();
  }

  // Other
  Future<TUIValueCallBack<TUILiveInfo>> fetchLiveInfo(String roomId) {
    return _roomManager.fetchLiveInfo(roomId);
  }

  Future<TUIValueCallBack<TUIUserInfo>> getUserInfo(String userId) {
    return _userManager.getUserInfo(userId);
  }

  Future<TUIActionCallback> onDisableSendingMessageBtnClicked(String userId, bool isDisable) {
    return _userManager.onDisableSendingMessageBtnClicked(userId, isDisable);
  }

  Future<TUIActionCallback> onKickedOutBtnClicked(String userId) {
    return _userManager.onKickedOutBtnClicked(userId);
  }

  Future<TUIActionCallback> onLockMediaStatusBtnClicked(String userId, TUISeatLockParams params) {
    return _coGuestManager.onLockMediaStatusBtnClicked(userId, params);
  }

  bool isCoGuesting() {
    return coreCoGuestState.seatList.value
        .where((user) => user.userId.isNotEmpty && user.userId != coreUserState.selfInfo.userId)
        .isNotEmpty;
  }

  void updateVideoQuality(TUIVideoQuality videoQuality) {
    return _mediaManager.updateVideoQuality(videoQuality);
  }

  Future<TUIValueCallBack<List<TUIVideoQuality>>> getMultiPlaybackQuality(String roomId) {
    return _mediaManager.getMultiPlaybackQuality(roomId);
  }

  void switchPlaybackQuality(TUIVideoQuality videoQuality) {
    return _mediaManager.switchPlaybackQuality(videoQuality);
  }

  void setAudioPlayoutVolume(int volume) {
    return _mediaManager.setAudioPlayoutVolume(volume);
  }

  void pauseByAudience() {
    return _mediaManager.pauseByAudience();
  }

  void resumeByAudience() {
    return _mediaManager.resumeByAudience();
  }
}

extension LiveStreamManagerWithAudience on LiveStreamManager {
  void onJoinLive(TUILiveInfo liveInfo) {
    _mediaManager.onJoinLive(liveInfo);
    return _roomManager.onJoinLive(liveInfo);
  }

  void onLeaveLive() {
    _roomManager.onLeaveLive();
    _userManager.onLeaveLive();
    _mediaManager.onLeaveLive();
  }

  void onStartRequestIntraRoomConnection() {
    _coGuestManager.onStartRequestIntraRoomConnection();
  }

  void onRequestIntraRoomConnectionFailed() {
    _coGuestManager.onRequestIntraRoomConnectionFailed();
  }

  void onStartCancelIntraRoomConnection() {
    _coGuestManager.onStartCancelIntraRoomConnection();
  }

  void onCancelIntraRoomConnection() {
    _coGuestManager.onCancelIntraRoomConnection();
  }

  void onCoGuestStatusChanged(CoGuestStatus status) {
    // TODO: krab needs to confirm whether needs changeVideoEncParams logic here
  }
}

extension LiveStreamManagerWithFloatWindow on LiveStreamManager {

  void enablePipMode(bool enable) {
    _floatWindowManager.enablePipMode(enable);
  }

  String buildEnablePipJsonParams(bool enable, String roomId) {
    return _floatWindowManager.buildEnablePipJsonParams(enable, roomId);
  }

  void setFloatWindowMode(FloatWindowMode mode) {
    LSFloatWindowState floatWindowState = _floatWindowManager.floatWindowState;
    if (floatWindowState.floatWindowMode is ValueNotifier<FloatWindowMode>) {
      ValueNotifier<FloatWindowMode> floatWindowMode =
          floatWindowState.floatWindowMode as ValueNotifier<FloatWindowMode>;
      floatWindowMode.value = mode;
    }
  }
}

extension LiveStreamManagerWithTools on LiveStreamManager {
  // State
  LSRoomState get roomState => _roomManager.roomState;

  LSUserState get userState => _userManager.userState;

  LSMediaState get mediaState => _mediaManager.mediaState;

  LSCoGuestState get coGuestState => _coGuestManager.coGuestState;

  LSCoHostState get coHostState => _coHostManager.coHostState;

  LSBattleState get battleState => _battleManager.battleState;

  LSFloatWindowState get floatWindowState => _floatWindowManager.floatWindowState;

  // Manager
  CoHostManager get coHostManager => _coHostManager;

  BattleManager get battleManager => _battleManager;

  MediaManager get mediaManager => _mediaManager;

  // Other
  String getDefaultRoomName() {
    return _roomManager.getDefaultRoomName();
  }
}

extension LiveStreamManagerWithCoreStateProvider on LiveStreamManager {
  RoomState get coreRoomState => _context.coreRoomState;

  UserState get coreUserState => _context.coreUserState;

  MediaState get coreMediaState => _context.coreMediaState;

  CoGuestState get coreCoGuestState => _context.coreCoGuestState;

  CoHostState get coreCoHostState => _context.coreCoHostState;

  BattleState get coreBattleState => _context.coreBattleState;
}

extension on LiveStreamManager {
  void _init() {
    service = LiveStreamService();
    toastSubject = StreamController.broadcast();
    likeSubject = StreamController.broadcast();
    kickedOutSubject = StreamController.broadcast();

    _roomManager = RoomManager();
    _userManager = UserManager();
    _mediaManager = MediaManager();
    _coGuestManager = CoGuestManager();
    _coHostManager = CoHostManager();
    _battleManager = BattleManager();
    _floatWindowManager = FloatWindowManager();
    _roomEngineObserver = RoomEngineObserver();
    _liveListObserver = LiveListObserver();
    liveStreamObserver = LiveStreamObserver();
    battleManagerObserver = BattleManagerObserver();
    _liveLayoutObserver = LiveLayoutObserver();

    _context = Context(
        service: service,
        toastSubject: WeakReference(toastSubject),
        likeSubject: WeakReference(likeSubject),
        kickedOutSubject: WeakReference(kickedOutSubject),
        roomManager: WeakReference(_roomManager),
        userManager: WeakReference(_userManager),
        mediaManager: WeakReference(_mediaManager),
        coGuestManager: WeakReference(_coGuestManager),
        coHostManager: WeakReference(_coHostManager),
        battleManager: WeakReference(_battleManager),
        roomEngineObserver: WeakReference(_roomEngineObserver),
        liveListObserver: WeakReference(_liveListObserver),
        liveStreamObserver: WeakReference(liveStreamObserver),
        battleManagerObserver: WeakReference(battleManagerObserver),
        liveLayoutObserver: WeakReference(_liveLayoutObserver),
        provider: WeakReference(provider));

    _roomManager.init(_context);
    _userManager.init(_context);
    _mediaManager.init(_context);
    _coGuestManager.init(_context);
    _coHostManager.init(_context);
    _battleManager.init(_context);
    _floatWindowManager.init(_context);

    _roomEngineObserver.init(_context);
    _liveListObserver.init(_context);
    liveStreamObserver.init(_context);
    battleManagerObserver.init(_context);
    _liveLayoutObserver.init(_context);

    service.addEngineObserver(_roomEngineObserver);
    service.addLiveListManagerObserver(_liveListObserver);
    service.addLiveLayoutObserver(_liveLayoutObserver);
  }

  void _unInit() {
    _roomManager.dispose();
    _userManager.dispose();
    _mediaManager.dispose();
    _coGuestManager.dispose();
    _coHostManager.dispose();
    _battleManager.dispose();
    _floatWindowManager.dispose();

    service.removeEngineObserver(_roomEngineObserver);
    service.removeLiveListManagerObserver(_liveListObserver);
    service.removeLiveLayoutObserver(_liveLayoutObserver);
    if (!toastSubject.isClosed) {
      toastSubject.close();
    }
    if (!likeSubject.isClosed) {
      likeSubject.close();
    }
    if (!kickedOutSubject.isClosed) {
      kickedOutSubject.close();
    }
  }
}

class Context {
  late final LiveStreamService service;
  late final WeakReference<StreamController<String>> toastSubject;
  late final WeakReference<StreamController<void>> likeSubject;
  late final WeakReference<StreamController<void>> kickedOutSubject;

  late final WeakReference<RoomManager> roomManager;
  late final WeakReference<UserManager> userManager;
  late final WeakReference<MediaManager> mediaManager;
  late final WeakReference<CoGuestManager> coGuestManager;
  late final WeakReference<CoHostManager> coHostManager;
  late final WeakReference<BattleManager> battleManager;

  late final WeakReference<RoomEngineObserver> roomEngineObserver;
  late final WeakReference<LiveListObserver> liveListObserver;
  late final WeakReference<LiveStreamObserver> liveStreamObserver;
  late final WeakReference<BattleManagerObserver> battleManagerObserver;
  late final WeakReference<LiveLayoutObserver> liveLayoutObserver;
  late final WeakReference<CoreStateProvider> provider;

  Context(
      {required this.service,
      required this.toastSubject,
      required this.likeSubject,
      required this.kickedOutSubject,
      required this.roomManager,
      required this.userManager,
      required this.mediaManager,
      required this.coGuestManager,
      required this.coHostManager,
      required this.battleManager,
      required this.roomEngineObserver,
      required this.liveListObserver,
      required this.liveStreamObserver,
      required this.battleManagerObserver,
      required this.liveLayoutObserver,
      required this.provider});

  RoomState get coreRoomState => provider.target?.getCoreState().roomState ?? RoomState();

  UserState get coreUserState => provider.target?.getCoreState().userState ?? UserState();

  MediaState get coreMediaState => provider.target?.getCoreState().mediaState ?? MediaState();

  CoGuestState get coreCoGuestState => provider.target?.getCoreState().coGuestState ?? CoGuestState();

  CoHostState get coreCoHostState => provider.target?.getCoreState().coHostState ?? CoHostState();

  BattleState get coreBattleState => provider.target?.getCoreState().battleState ?? BattleState();

  LayoutState get coreLayoutState => provider.target?.getCoreState().layoutState ?? LayoutState();
}
